prov/rxd/src/rxd.h

```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
+	uint64_t done;
 	uint32_t seg_no;
 	uint32_t window;
+	uint32_t next_start;
 	uint64_t retry_time;
 	uint8_t retry_cnt;
 
 	struct dlist_entry entry;
 	struct dlist_entry pkt_list;
 
-	uint8_t op_type;
 	struct ofi_op_hdr op_hdr;
+	struct iovec iov[RXD_IOV_LIMIT];
+	uint8_t iov_count;
+};
+DECLARE_FREESTACK(struct rxd_x_entry, rxd_tx_fs);
+DECLARE_FREESTACK(struct rxd_x_entry, rxd_rx_fs);
 
-	union {
-		struct {
-			struct fi_msg msg;
-			struct iovec msg_iov[RXD_IOV_LIMIT];
-		} msg;
+#define RXD_RETRY	(1 << 0)
+#define RXD_LAST	(1 << 1)
 
-		struct {
-			struct fi_msg_tagged tmsg;
-			struct iovec msg_iov[RXD_IOV_LIMIT];
-		} tmsg;
+struct rxd_data_hdr {
+	uint32_t tx_id;
+	uint32_t rx_id;
+	uint32_t key;
+	uint32_t seg_no;
+	uint32_t flags;
+	fi_addr_t peer;
+};
 
-		struct {
-			struct fi_msg_rma msg;
-			struct iovec src_iov[RXD_IOV_LIMIT];
-			struct fi_rma_iov dst_iov[RXD_IOV_LIMIT];
-		} write;
+struct rxd_ctrl_hdr {
+	uint32_t tx_id;
+	uint32_t rx_id;
+	uint32_t key;
 
-		struct {
-			struct fi_msg_rma msg;
-			struct fi_rma_iov src_iov[RXD_IOV_LIMIT];
-			struct iovec dst_iov[RXD_IOV_LIMIT];
-		} read_req;
+	fi_addr_t peer;
+	uint8_t flags;
+	enum rxd_msg_type type;
+	uint32_t window;
+	size_t size;
+	uint64_t data;
+};
 
+struct rxd_pkt {
+	union {
+		struct {
+			uint32_t resv;//always = ~0
+			struct rxd_ctrl_hdr ctrl_hdr;
+			uint8_t source[];
+		};
 		struct {
-			uint64_t peer_msg_id;
-			uint8_t iov_count;
-			struct iovec src_iov[RXD_IOV_LIMIT];
-		} read_rsp;
+			struct rxd_data_hdr data_hdr;
+			uint8_t data[];
+		};
 	};
 };
-DECLARE_FREESTACK(struct rxd_tx_entry, rxd_tx_entry_fs);
 
-struct rxd_recv_entry {
-	struct dlist_entry entry;
-	struct fi_msg msg;
-	uint64_t flags;
-	struct iovec iov[RXD_IOV_LIMIT];
-	void *desc[RXD_IOV_LIMIT];
-};
-DECLARE_FREESTACK(struct rxd_recv_entry, rxd_recv_fs);
+#define RXD_CTRL_PKT_SIZE (sizeof(struct rxd_ctrl_hdr) + RXD_NAME_LENGTH)
 
-struct rxd_trecv_entry {
+struct rxd_pkt_entry {
 	struct dlist_entry entry;
-	struct fi_msg_tagged msg;
-	uint64_t flags;
-	struct rxd_rx_entry *rx_entry;
-	struct iovec iov[RXD_IOV_LIMIT];
-	void *desc[RXD_IOV_LIMIT];
-};
-DECLARE_FREESTACK(struct rxd_trecv_entry, rxd_trecv_fs);
-
-struct rxd_pkt_data_start {
-	struct ofi_ctrl_hdr ctrl;
-	struct ofi_op_hdr op;
-	char data[];
-};
-
-struct rxd_pkt_data {
-	struct ofi_ctrl_hdr ctrl;
-	char data[];
-};
-
-#define RXD_PKT_FIRST	(1 << 0)
-#define RXD_PKT_LAST	(1 << 1)
-#define RXD_LOCAL_COMP	(1 << 2)
-#define RXD_REMOTE_ACK	(1 << 3)
-#define RXD_NOT_ACKED	RXD_REMOTE_ACK
-
-struct rxd_pkt_meta {
+	size_t pkt_size;
 	struct fi_context context;
-	struct dlist_entry entry;
-	struct rxd_tx_entry *tx_entry;
-	struct rxd_ep *ep;
 	struct fid_mr *mr;
-	int flags;
-
-	/* TODO: use iov and remove data copies */
-	char pkt_data[]; /* rxd_pkt_data*, followed by data */
+	fi_addr_t peer;
+	struct rxd_pkt pkt;
 };
 
+static inline int rxd_is_data_pkt(struct rxd_pkt_entry *pkt_entry)
+{
+	return (pkt_entry->pkt.resv != ~0);
```

Comment:
use a type field, not a reserved field that isn't actually reserved and must be set to a specific value.